lib/mongodb.js

```diff
@@ -1870,30 +1873,71 @@ MongoDB.prototype.ping = function(cb) {
   }
 };
 
+// Determine if a property must have an ObjectID value
+function mustBeObjectID(propDef) {
+  if (!propDef) return false;
+  if (propDef.mongodb && ObjectIdTypeRegex.test(propDef.mongodb.dataType)) return true;
+  return false;
+}
+
+// Determine if strictObjectIDCoercion should be enabled
+function enableStrictObjectIDCoercion(modelCtor, options) {
+  const settings = modelCtor.settings;
+  return (settings && settings.strictObjectIDCoercion) ||
+    (modelCtor.model && modelCtor.model.getConnector().settings.strictObjectIDCoercion) ||
+    options &&
+    options.strictObjectIDCoercion;
+}
+
+function coerceToObjectId(modelCtor, propDef, propValue) {
+  if (mustBeObjectID(propDef)) {
+    if (isObjectIDProperty(modelCtor, propDef, propValue)) {
+      return ObjectID(propValue);
+    } else {
+      throw new Error(`${propValue} is not an ObjectID string`);
+    }
+  } else if (enableStrictObjectIDCoercion(modelCtor)) {
+    if (isObjectIDProperty(modelCtor, propDef, propValue)) {
+      return ObjectID(propValue);
+    }
+  } else if (ObjectIdRegex.test(propValue)) {
+    return ObjectID(propValue);
+  }
+  return propValue;
+}
+
 /**
  * Check whether the property is an ObjectID (or Array thereof)
  *
  */
-MongoDB.prototype.isObjectIDProperty = function(modelName, prop, value, options) {
+function isObjectIDProperty(modelCtor, propDef, value, options) {
```

Comment:
What's the difference between `isObjectIDProperty` and `mustBeObjectID`? Can we merge those two methods into a single one?

Reply:
`mustBeObjectID` determines whether a property's value must be `ObjectID`.  It is like specifying the data type.

`isObjectIDProperty` checks if the property value is an `ObjectID`. It is like checking the data type. This is how it does the determination:

- If the value doesn't look like an ObjectID -> `false`
- If the value looks like an ObjectID and `strictObjectIDCoercion` is set to `false` (default) and `mongodb.dataType` is not set -> `true`
- If the value looks like an ObjectID and `strictObjectIDCoercion` is set to `true` and `mongodb.dataType` is not set -> `false`
- If the value looks like an ObjectID and `strictObjectIDCoercion` is set to `true` and `mongodb.dataType` is set to "ObjectID" -> true

I actually created these two functions out of a single function to add clarity to the flow. I am not in favor on re-merging them, since they serve different purposes and can be used independently in other places.

Reply:
We are actually checking the data type in `isObjectIDProperty` (see line 1922) so maybe we can call `mustBeObjectID` for it there. That function is called only once on line 1893 btw. My 2 cents is to keep both and perhaps rename `mustBeObjectID` to a more descriptive name.

Reply:
> so maybe we can call mustBeObjectID for it there

Thanks, done.

Reply:
I find this problematic. IMO we should not make any distinction between properties defined as `{type: 'ObjectID'}` and `{type: 'string', mongodb: {dataType: 'ObjectID'}}`.

In both cases, the property should be considered by the connector as having ObjectID type.

My main concern is that by having two functions, it's easy to use a wrong one and introduce situation where certain code paths don't respect `mongodb: {dataType: 'ObjectID'}` setting.

Reply:
Agreed.

Reply:
I get it now too. Yea we should aim at having unified logic for those code paths 👍 .

Reply:
@bajtos @hacksparrow can we create a follow-up story for this?

Reply:
I have fixed this one.